### PR TITLE
LF-4788 (Backend) Create GET endpoint to fetch individual IP from Ensemble

### PR DESCRIPTION
**Pull request by kathyavini**
**Description**

This is the backend portion of the GET endpoint that will ultimately receive and pass on the details for a specific irrigation prescription from Ensemble. To keep the number of touched files manageable, I will open the frontend in a separate PR. In that PR, the frontend mocks will be removed and RTK query calls to this endpoint made in their place.

The content of endpoint is really only three things:
1. Receiving data OR (with `shouldSend = false`) replicating the mocking currently done by the frontend
2. Tiny bit of format massaging (the units of the weather forecast, which as far as we know right now is only the ˚C → C conversion). The number of units to convert _might_ be expanded after we get the real data back, we'll want to double-check that
3. The water consumption calculation for URI and VRI

You can test the endpoint in Postman with the following GET

```
/irrigation_prescriptions/{{id}}?shouldSend=false
```

`id` can be any integer; odd ones will return VRI prescriptions and even ones will return URI prescriptions.

Some implementation notes:
- For simplicity, I did not use the addon partner construct of the other irrigation prescription GET (controller → addonPartner service → Ensemble service), but followed the format of all the other controllers currently calling Ensemble (controller → Ensemble service).
- The mocks are the same ones that can be see on the task details view (the ones @Duncan-Brain created in #3744)
- just as an extra sanity check, I arranged the application depths so that they are equivalent between the URI and VRI mock, so the water consumption should also be approximately comparable

Jira link: https://lite-farm.atlassian.net/browse/LF-4788

**Type of change**

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

**How Has This Been Tested?**

- [ ] Passes test case
- [ ] UI components visually reviewed on desktop view
- [x] UI components visually reviewed on mobile view
- [x] Other (please explain)

Tested in Postman as described above and alongside frontend branch (WIP but to be opened shortly)

**Checklist:**

- [x] I have commented my code, particularly in hard-to-understand areas
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes
- [x] The precommit and linting ran successfully
- [ ] I have added or updated language tags for text that's part of the UI
- [ ] I have ordered translation keys alphabetically (optional: run `pnpm i18n` to help with this)
- [x] I have added [the GNU General Public License](https://lite-farm.atlassian.net/l/cp/BT0Dd7WW) to all new files

### CODE REVIEW

File: packages/api/src/util/ensembleService.ts [hunk 1]
```diff
@@ -46,18 +53,17 @@ const getAddonPartnerId = async (): Promise<AddonPartner['id']> => {
...
- * Retrieves the external organisation IDs for a specific farm and partner.
+ * Retrieves the Ensemble addon partner ids for the given farm, and throws an error if the farm is not connected to Ensemble.
  *
  * @param farmId - The ID of the farm to retrieve external organisation IDs for.
- * @param addonPartnerId - The ID of addOnPartner for whose endpoint the ids are compatible with.
  * @returns A promise that resolves to the organisation IDs for the given farm and partner.
  * @throws Not found error as we expect that the farms addon partner ids exist.
  */
-const getExternalOrganisationIds = async (
+const getFarmEnsembleAddonIds = async (
```
Comment (kathyavini) [inline packages/api/src/util/ensembleService.ts:62]: @Duncan-Brain I mentioned this in the other PR but I think I forgot to follow-up! The JSDoc for the helper function above was still not right, and the error message that this throws only makes sense in the case of the passed partner being Ensemble (so I assumed you didn't mean to require passing it). I think this is consistent with how you used it though; let me know if not.

File: packages/api/src/util/generateMockPrescriptionDetails.ts [hunk 2]
```diff
@@ -0,0 +1,165 @@
...
+  const mockPivot = {
+    center: getCentroidOfPolygon(locationPolygon) ?? locationPolygon[0],
+    radius: pivotRadius,
+  };
+
+  const commonMockData = {
+    id: ip_id,
+    location_id: locations[0].location_id,
+    management_plan_id: managementPlans[0]?.management_plan_id ?? null,
+    recommended_start_datetime: getDateTimeFromDayOfMonth(ip_id).toISOString(),
```
Comment (Duncan-Brain) [inline packages/api/src/util/generateMockPrescriptionDetails.ts:68]: Wondering if we should align both mocks shared data using the same function for location_id, management_plan_id, recommended_start_time. 

Not sure if there are any actual implications for testing task creation or whatever but could be handy.

File: packages/api/src/util/mockAddonPartner.ts [hunk 3]
```diff
@@ -39,11 +39,11 @@ export const getIrrigationPrescriptions = async ({
 }: fakeIrrigationPrescriptionsProps): Promise<ExternalIrrigationPrescription[]> => {
   const PRESCRIPTION_CONFIG = [
     {
-      id: 1,
+      id: new Date(startTime ?? Date.now()).getUTCDate(),
```
Comment (Duncan-Brain) [inline packages/api/src/util/mockAddonPartner.ts:42]: Since this only updates one new one per day, it might still be worth showing Denis the correct column in `irrigation_task` table in beta db to delete the id. He should have access to that.

**Review by Duncan-Brain — COMMENTED** [review]
Left a review, no major issues that I can see as is. It will be more clear when the frontend is updated. Only thing I haven't looked into hard enough is the management_plan_id interaction on frontend.

File: packages/api/src/controllers/irrigationPrescriptionController.ts [hunk 4]
```diff
@@ -47,6 +52,36 @@ const irrigationPrescriptionController = {
...
     };
   },
+
+  getPrescriptionDetails() {
+    return async (
+      req: LiteFarmRequest<PrescriptionDetailsQueryParams, PrescriptionDetailsRouteParams>,
+      res: Response,
+    ) => {
+      const { farm_id } = req.headers;
+      const { ip_id } = req.params;
```
[the remark this answers is not in the document]
Comment (kathyavini) [reply packages/api/src/controllers/irrigationPrescriptionController.ts]: Good point! We use "ip" so much verbally I forgot it's not a codebase convention at all. I like it a lot more written out too ❤️

File: packages/api/src/util/ensembleService.ts [hunk 5]
```diff
@@ -161,6 +165,127 @@ export async function mockFetchIrrigationPrescriptionsFromEnsemble(org_pk: numbe
...
+  return {
+    ...rest,
+    metadata: {
+      weather_forecast: mappedWeatherForecast,
+    },
+  };
+};
+
+/**
+ * Calculates water consumption for Uniform Rate Irrigation (URI).
```
[the remark this answers is not in the document]
Comment (kathyavini) [reply packages/api/src/util/ensembleService.ts:258]: It's in the `@returns` line of the JSDoc.

File: packages/api/src/util/ensembleService.ts [hunk 6]
```diff
@@ -161,6 +165,127 @@ export async function mockFetchIrrigationPrescriptionsFromEnsemble(org_pk: numbe
...
+ * @param {EsciReturnedPrescriptionDetails['prescription']} prescription - The prescription object containing VRI data.
+ * @returns {number} The calculated water consumption in liters.
+ */
+const calculateVRIWaterConsumption = (
+  prescription: EsciReturnedPrescriptionDetails['prescription'],
+): number => {
+  const prescriptionZones: VriPrescriptionData[] = prescription.vriData!.zones;
+
+  return prescriptionZones.reduce((acc, zone) => {
+    const zoneAreaM2 = getAreaOfPolygon(zone.grid_points) ?? 0;
```
[the remark this answers is not in the document]
Comment (kathyavini) [reply packages/api/src/util/ensembleService.ts:285]: Could you elaborate on "allowing it to"? Do you mean you would rather throw and return an error from the endpoint if the calculation fails, rather than just ignoring the zone in the water calculation?

Since ESci is ultimately responsible for defining the zones, conceivably they could pass something that's not a valid polygon. I think the plus side of returning that IP with faulty zone nonetheless is that it will be immediately apparent on the map (it won't render) which is better feedback to the user than an error message about the area calculation.

I'm not sure if that addresses what you're asking, though.

File: packages/api/src/util/generateMockPrescriptionDetails.ts (hunk 2 above)
Comment (kathyavini) [reply packages/api/src/util/generateMockPrescriptionDetails.ts]: It should be clearer on the next PR (the frontend integration), but they are actually already aligned for the full task flow; I set them up specifically for that. Let's revisit after that PR if it's still not clear.

File: packages/api/tests/irrigation_prescription.test.ts [hunk 7]
```diff
@@ -178,4 +205,130 @@ describe('Get Irrigation Prescription Tests', () => {
...
+              temperature_unit: 'C',
+            },
+          },
+        });
+      });
+    });
+  });
+
+  describe('Water consumption calculuation tests', () => {
+    test('API should calculate water consumption for a VRI prescription', async () => {
```
[the remark this answers is not in the document]
Comment (kathyavini) [reply packages/api/tests/irrigation_prescription.test.ts:257]: As basic as they are, they are a part of the endpoint's logic (in fact, their only logic) and I wouldn't feel comfortable omitting them entirely. But I agree that if we had unit tests, instead of just API tests, that would be a better home for them. This is more of a compromise working within the existing test structure; maybe that's what feels out of place.

Do we aim to cover each of our throw statements with a test case? This hasn't been the pattern for recent endpoints. Let me also defer to @SayakaOno here, though, because I trust her testing strategy 😂 Are there any in particular you thought should warrant a test case?

File: packages/api/src/util/ensembleService.types.ts [hunk 8]
```diff
@@ -120,3 +120,76 @@ export function isExternalIrrigationPrescriptionArray(
...
+  application_depth_unit: string;
+}
+
+export type VriPrescriptionData = UriPrescriptionData & {
+  grid_points: Point[];
+};
+
+export type LiteFarmWeatherUnits = 'mm' | 'mm/h' | 'C' | 'm/s';
+
+export type IrrigationPrescriptionDetails = Omit<EsciReturnedPrescriptionDetails, 'metadata'> & {
```
[the remark this answers is not in the document]
Comment (kathyavini) [reply packages/api/src/util/ensembleService.types.ts]: Sounds good 👍 I wasn't liking that 'Omit' too much on second look either 😂

File: packages/api/src/util/ensembleService.ts [hunk 9]
```diff
@@ -161,6 +165,127 @@ export async function mockFetchIrrigationPrescriptionsFromEnsemble(org_pk: numbe
...
+ *
+ * @param {EsciReturnedPrescriptionDetails} prescription - The prescription details from Ensemble API.
+ * @returns {EsciReturnedPrescriptionDetails} The prescription with units mapped to LiteFarm format.
+ */
+const mapEnsembleUnitsToLiteFarmUnits = (prescription: EsciReturnedPrescriptionDetails) => {
+  const { metadata, ...rest } = prescription;
+
+  const mapWeatherUnit = (unit: EsciWeatherUnits): LiteFarmWeatherUnits => {
+    if (unit === '˚C') return 'C';
+    return unit as LiteFarmWeatherUnits;
```
[the remark this answers is not in the document]
Comment (kathyavini) [reply packages/api/src/util/ensembleService.ts]: Thanks! Didn't catch that.

It's not a type guard though, just a mapping function where we can put in any unit conversions we need.

File: packages/api/src/util/ensembleService.ts [hunk 10]
```diff
@@ -161,6 +165,127 @@ export async function mockFetchIrrigationPrescriptionsFromEnsemble(org_pk: numbe
...
+/**
+ * Maps units from Ensemble API format to LiteFarm format.
+ *
+ * @param {EsciReturnedPrescriptionDetails} prescription - The prescription details from Ensemble API.
+ * @returns {EsciReturnedPrescriptionDetails} The prescription with units mapped to LiteFarm format.
+ */
+const mapEnsembleUnitsToLiteFarmUnits = (prescription: EsciReturnedPrescriptionDetails) => {
+  const { metadata, ...rest } = prescription;
+
+  const mapWeatherUnit = (unit: EsciWeatherUnits): LiteFarmWeatherUnits => {
```
[the remark this answers is not in the document]
Comment (kathyavini) [reply packages/api/src/util/ensembleService.ts:237]: There are no type guards in this mapping function (no checks on acceptable measurements), it's just doing what `ESCI_TO_CONVERT_UNITS_MAP` does in the other file, but in a way that satisfies the type checker.

I do want some kind of non-erroring check -- now I'm thinking of omitting those properties that don't come in our expected unit -- but I want to double-check that the KPIs are set up to handle that. I have a note in our tech daily doc to talk about this!

File: packages/api/src/util/ensembleService.ts (hunk 5 above)
[the remark this answers is not in the document]
Comment (Duncan-Brain) [reply packages/api/src/util/ensembleService.ts:258]: Guess I missed it! 🙏

File: packages/api/src/util/ensembleService.ts [hunk 11]
```diff
@@ -161,6 +165,127 @@ export async function mockFetchIrrigationPrescriptionsFromEnsemble(org_pk: numbe
...
+  }
+  if (prescriptionFarmRecord.farm_id !== farm_id) {
+    throw customError(`Irrigation prescription ${ip_id} belongs to a different farm`, 403);
+  }
+
+  // Transform prescription data to LiteFarm format and validate details
+  const mappedPrescription = mapEnsembleUnitsToLiteFarmUnits(irrigationPrescription);
+  const prescriptionDetails = mappedPrescription.prescription;
+  if (!prescriptionDetails) {
+    throw customError('Prescription data is missing', 500);
```
[the remark this answers is not in the document]
Comment (Duncan-Brain) [reply packages/api/src/util/ensembleService.ts:212]: So if we can check for errors prior to other code it might be better to return early. For example we could know about this error before we call `LocationModel`.  I think a type guard or other formatting check that the external response is as expected immediately after we receive it, is valuable.

so maybe like this is what I might be expecting:
```
// get response
// check response format
// check farm access
// format + derive data
// send to fronted
```

I don't think there is an issue with being strict in what we expect to work. But I agree it could potentially be valuable to accept more null values if we determine the core data is correct.

File: packages/api/src/util/ensembleService.ts (hunk 6 above)
[the remark this answers is not in the document]
Comment (Duncan-Brain) [reply packages/api/src/util/ensembleService.ts:285]: I am not confident that an unclosed polygon could potentially fail some function but render correctly. It seems possible that a map could work with an unclosed polygon and a function could not.

Even if not, I also think both error options kind of stink haha. If only 1/3 vri zones fails then you will get some number (which may not be double checked). So to me `0` could lead to false (and stored in task db) estimations. 

I haven't thought it through much but I might prefer a `null` over a misleading value. 

We could do a check on the response from the api and also check polygons correctness.

File: packages/api/src/util/mockAddonPartner.ts (hunk 3 above)
Comment (Duncan-Brain) [reply packages/api/src/util/mockAddonPartner.ts:42]: Ah yes, I always forget about other farms and task deleting. Thanks! .. I thought for sure he had access.

File: packages/api/tests/irrigation_prescription.test.ts (hunk 7 above)
[the remark this answers is not in the document]
Comment (Duncan-Brain) [reply packages/api/tests/irrigation_prescription.test.ts:343]: Testing the logic: Yeah for sure but I am thinking for those sorts of tests it is useful to test edge cases in logic where you think it might fail due to the complexity (setting something to zero, undefined, etc). 

It should also be possible to import those functions directly into this test and test them directly here without involving the endpoint.

My goal is to cover as many `throw` statements as possible (animals I worked hard on that because it was a new module), but its not always practical with the difficulty of our tests. So if not practical then critical ones I think we normally do. Because our tests are so complex I often fail at my side goal of increasing code coverage. 

I think the most valuable tests for this would be :
```
// Users should successfully GET  ✅ 

// Returns an error if eternal api does not return data ❓ 

// Will not pass data from another farm ❓ 
```

I am not strict on tests or anything .. just what I am seeing..

File: packages/api/src/util/ensembleService.ts (hunk 9 above)
[the remark this answers is not in the document]
Comment (Duncan-Brain) [reply packages/api/src/util/ensembleService.ts]: Yeah I use type guard and narrowing loosely haah -- will work on that. What would this be called then a type transformer 🤖  haha

File: packages/api/src/util/ensembleService.types.ts [hunk 12]
```diff
@@ -120,3 +120,76 @@ export function isExternalIrrigationPrescriptionArray(
...
+
+  metadata: {
+    weather_forecast: {
+      temperature: number;
+      temperature_unit: EsciWeatherUnits;
+      wind_speed: number;
+      wind_speed_unit: EsciWeatherUnits;
+      cumulative_rainfall: number;
+      cumulative_rainfall_unit: EsciWeatherUnits;
+      et_rate: number;
```
[the remark this answers is not in the document]
Comment (Duncan-Brain) [reply packages/api/src/util/ensembleService.types.ts]: 🙏

File: packages/api/src/util/generateMockPrescriptionDetails.ts (hunk 2 above)
Comment (Duncan-Brain) [reply packages/api/src/util/generateMockPrescriptionDetails.ts]: Well in the other mock I ignore `management_plan_id entirely` and I think we calculate recommended start time differently. Location looks the same.  

I think what I was thinking is maybe since we are reusing stuff, and they are menat to be identical, if we should make a function to return those 3 things or not and use it in both.

File: packages/api/tests/irrigation_prescription.test.ts (hunk 7 above)
[the remark this answers is not in the document]
Comment (kathyavini) [reply packages/api/tests/irrigation_prescription.test.ts:343]: I like that selection of tests for throws! I might also add the metadata in wrong units and the case of bad zones for water calculation, after we discuss with Loïc how we're going to flag that.

[the remark this answers is not in the document]
Comment (SayakaOno) [reply packages/api/tests/irrigation_prescription.test.ts:343]: Looking at this test case, it seems like it’s mainly checking whether the code to calculate water consumption for a VRI prescription runs, rather than verifying the correctness of the calculation. So, this test looks good to me!

For testing the actual calculation logic, I think unit tests in a separate file would be preferable!

Ah, I was answering a different question...😂 
It would be nice to cover as many cases as possible, but with limited time, there’s always some compromise... Personally, I wouldn't aim to cover every possible case, but rather focus on what I actually need to achieve with the tests. (Sorry for the useless input 😇) The test cases Duncan suggested make sense to me too!

**Comment by kathyavini** [comment]
This was discussed synchronously today, but just to update:
- We'll decide on the core data and the error states we want to show in the UI with Loïc in dev-design on Wednesday
- For the time being, IPs on hold in favour of finishing sensors + as much of soil sampling task as possible this week for bashing on Friday

File: packages/api/src/util/ensembleService.ts [hunk 13]
```diff
@@ -161,6 +165,127 @@ export async function mockFetchIrrigationPrescriptionsFromEnsemble(org_pk: numbe
...
+  const prescriptionFarmRecord = await LocationModel.getFarmIdByLocationId(
+    irrigationPrescription.location_id,
+  );
+  if (!prescriptionFarmRecord) {
+    throw customError(`location_id on IP ${ip_id} does not exist`, 404);
+  }
+  if (prescriptionFarmRecord.farm_id !== farm_id) {
+    throw customError(`Irrigation prescription ${ip_id} belongs to a different farm`, 403);
+  }
+
```
[the remark this answers is not in the document]
Comment (kathyavini) [reply packages/api/src/util/ensembleService.ts:207]: According to our tech daily discussion, we'll be ignoring this one since Ensemble is not including our crop data at this stage of the project.

File: packages/api/tests/irrigation_prescription.test.ts (hunk 7 above)
[the remark this answers is not in the document]
Comment (kathyavini) [reply packages/api/tests/irrigation_prescription.test.ts:343]: Added the two throws above for now @Duncan-Brain; these tests might need adjust in the user story where we handle the error states (I'll link that Jira below) but they match the current controller flow.

**Comment by kathyavini** [comment]
There are still some open discussion points on this on this PR around the error states we want to flag and how.

However, now there is a separate user-story specifically for IP error states: [LF-4877](https://lite-farm.atlassian.net/browse/LF-4877) and I have moved our tech daily notes around this there; please also feel free to add in the discussion for that user story 🙏 

I think as we discussed a few weeks ago, for this ticket we'll keep pretty minimal on the error states and assume the happy path so we can wrap up the feature.

**Review by Duncan-Brain — APPROVED** [review]
